src/block_gmres.jl

```diff
@@ -0,0 +1,349 @@
+                      ldiv::Bool=false, restart::Bool=false, reorthogonalization::Bool=false,
+                      atol::T = √eps(T), rtol::T=√eps(T), itmax::Int=0,
+                      timemax::Float64=Inf, verbose::Int=0, history::Bool=false) where {T <: AbstractFloat, FC <: FloatOrComplex{T}, S <: AbstractMatrix{FC}}
+
+  # Timer
+  start_time = time_ns()
+  timemax_ns = 1e9 * timemax
+
+  n, m = size(A)
+  s, p = size(B)
+  m == n || error("System must be square")
+  n == s || error("Inconsistent problem size")
+  (verbose > 0) && @printf("BLOCK-GMRES: system of size %d with %d right-hand sides\n", n, p)
+
+  # Check M = Iₙ and N = Iₙ
+  MisI = (M === I)
+  NisI = (N === I)
+
+  # Check type consistency
+  eltype(A) == FC || @warn "eltype(A) ≠ $FC. This could lead to errors or additional allocations in operator-matrix products."
+  ktypeof(B) <: S || error("ktypeof(B) is not a subtype of $S")
+
+  # Set up workspace.
+  allocate_if(!MisI  , solver, :Q , S, n, p)
+  allocate_if(!NisI  , solver, :P , S, n, p)
+  allocate_if(restart, solver, :ΔX, S, n, p)
+  ΔX, X, W, V, Z = solver.ΔX, solver.X, solver.W, solver.V, solver.Z
+  R, stats = solver.R, solver.stats
+  warm_start = solver.warm_start
+  RNorms = stats.residuals
+  reset!(stats)
+  Q  = MisI ? W : solver.Q
+  R₀ = MisI ? W : solver.Q
+  Xr = restart ? ΔX : X
+
+  # FIX ME
+  Ψtmp = Ztmp = S(undef, p, p)
+  τ = S(undef, p)
+
+  # Coefficients for mul!
+  α = -one(FC)
+  β = one(FC)
+  γ = one(FC)
+
+  # Initial solution X₀.
+  X .= zero(FC)
+
+  # Initial residual R₀.
+  if warm_start
+    mul!(W, A, Δx)
+    W .= B .- W
+    restart && (X .+= ΔX)
+  else
+    W .= B
+  end
+  MisI || mulorldiv!(R₀, M, W, ldiv)  # R₀ = M(B - AX₀)
+  RNorm = norm(R₀)                    # ‖R₀‖_F   
+
+  history && push!(RNorms, RNorm)
+  ε = atol + rtol * RNorm
+
+  mem = length(V)  # Memory
+  npass = 0        # Number of pass
+
+  iter = 0        # Cumulative number of iterations
+  inner_iter = 0  # Number of iterations in a pass
+
+  itmax == 0 && (itmax = 2*div(n,p))
+  inner_itmax = itmax
+
+  (verbose > 0) && @printf("%5s  %5s  %7s  %5s\n", "pass", "k", "‖Rₖ‖", "timer")
+  kdisplay(iter, verbose) && @printf("%5d  %5d  %7.1e  %.2fs\n", npass, iter, rNorm, ktimer(start_time))
+
+  # Stopping criterion
+  solved = rNorm ≤ ε
+  tired = iter ≥ itmax
+  inner_tired = inner_iter ≥ inner_itmax
+  status = "unknown"
+  overtimed = false
+
+  while !(solved || tired ||  overtimed)
+
+    # Initialize workspace.
+    nr = 0  # Number of blocks Ψᵢⱼ stored in Rₖ.
+    for i = 1 : mem
+      V[i] .= zero(FC)  # Orthogonal basis of Kₖ(MAN, MR₀).
+    end
+    for Ψ in R
+      block .= zero(FC)  # Upper triangular matrix Rₖ.
+    end
+    for block in Z
+      block .= zero(FC)  # Right-hand of the least squares problem min ‖Hₖ₊₁.ₖYₖ - ΓE₁‖₂.
+    end
+
+    if restart
+      Xr .= zero(FC)  # Xr === ΔX when restart is set to true
+      if npass ≥ 1
+        mul!(W, A, X)
+        W .= B .- W
```

Comment:
isn't it redundant with line 161?

Reply:
The first time is when you potentially warm-start the method (initial X0) and this time is when you restart the method after a fixed number of iterations.